### PR TITLE
Add "Pride Missiles" cosmetic option to randomize small missile expansion colors

### PR DESCRIPTION
**Pull request by steven11sjf**
A new `pride_missiles` cosmetic option allows the patcher to randomize each missile pack to red (default), orange, yellow, green, blue or purple/violet. 

I decided to randomize patches within ODR instead of randomizing in RDV. I think it'd be needlessly complex to add multiple color variants in the generator, and having it in a cosmetic option makes more sense to me for what it does. This means two patches off the same config will have different-colored missile expansions, but I think it's an acceptable trade-off. 

image
image
image
image
image

### CODE REVIEW

**Review by duncathan — CHANGES_REQUESTED** [review]
the normal method for doing something like this would be to generate a random number in Randovania, and pass that number to the patcher to use as the seed for an RNG object

~~also we need more flags~~

**Comment by steven11sjf** [comment]
> the normal method for doing something like this would be to generate a random number in Randovania, and pass that number to the patcher to use as the seed for an RNG object
> 
> ~also we need more flags~

Ah, I can add that too. Also yeah, though i'm not sure where we could put flags in the game. I guess we could do the banner in the new game UI?

**Comment by duncathan** [comment]
mostly I was thinking other flags for the missiles. rainbow missiles, trans missiles, nb missiles, etc

**Comment by henriquegemignani** [comment]
A more flexible way is to just expose all these different models directly and have rdv color each pickup directly however it wants.

**Comment by steven11sjf** [comment]
Can you point me towards where to do that? I assume it's in the generator which I haven't really looked into a ton before.

**Comment by henriquegemignani** [comment]
> Can you point me towards where to do that? I assume it's in the generator which I haven't really looked into a ton before.

re-looking to this PR, all you'd need to do in ODR is the new model data entries you added.
For rdv, you'd add some special casing similar to what you did here, but [here](https://github.com/randovania/randovania/blob/main/randovania/games/dread/exporter/patch_data_factory.py#L234-L247) instead.
Not very different to how we do the [shiny missiles for Prime 1](https://github.com/randovania/randovania/blob/main/randovania/games/prime1/exporter/patch_data_factory.py#L133-L138).

File: open_dread_rando/missile_color_patcher.py [hunk 1]
```diff
@@ -0,0 +1,43 @@
+from copy import deepcopy
+from enum import Enum
+
+from open_dread_rando.patcher_editor import PatcherEditor
+
+from mercury_engine_data_structures.formats.bsmat import Bsmat
+
+# model names for pride missiles
+class MissileVariant(Enum):
```
Comment (henriquegemignani) [inline open_dread_rando/missile_color_patcher.py:9]: I'm not sure enums are the best case for this. Could have a dataclass with the mat_name/shader_path/rgba fields, and a global dict of all instances

File: open_dread_rando/missile_color_patcher.py [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+    BLACK = ("item_missile__BK", [0.0, 0.0, 0.0, 1.0])
+    GRAY = ("item_missile__GY", [0.5, 0.5, 0.5, 1.0])
+
+
+    def __init__(self, path: str, color: list[float]):
+        self.mat_name = f"{path}_mp_fxhologram_01"
+        self.shader_path = f"actors/items/item_missiletank/models/imats/{self.mat_name}.bsmat"
+
+        if len(color) != 4:
+            raise ValueError(f"Missile Variant {self.name} has {len(color)} values for RGBA, 4 is expected.")
```
Comment (henriquegemignani) [inline open_dread_rando/missile_color_patcher.py:28]: Typing could be a `tuple[float, float, float, float]`, would indicate this better.

**Review by henriquegemignani — CHANGES_REQUESTED** [review]

**Pull request merged**